File: android/src/main/java/dev/eduardoroth/mediaplayer/MediaPlayerPlugin.java
package dev.eduardoroth.mediaplayer;

import android.util.DisplayMetrics;
import android.util.Rational;
import android.view.ViewGroup;
import androidx.fragment.app.FragmentContainerView;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import dev.eduardoroth.mediaplayer.models.AndroidOptions;
import dev.eduardoroth.mediaplayer.models.ExtraOptions;
import dev.eduardoroth.mediaplayer.models.PlacementOptions;
import dev.eduardoroth.mediaplayer.models.SubtitleOptions;
import java.util.Objects;
import org.json.JSONException;

@CapacitorPlugin(name = "MediaPlayer")
public class MediaPlayerPlugin extends Plugin {

    private MediaPlayer implementation;

    @Override
    public void load() {
        implementation = new MediaPlayer(bridge.getActivity());
        MediaPlayerNotificationCenter.init(bridge.getActivity());
        MediaPlayerNotificationCenter.listenNotifications(nextNotification ->
            notifyListeners(nextNotification.getEventName(), nextNotification.getData())
        );

        ViewGroup coordinatorLayout = (ViewGroup) bridge.getActivity().findViewById(R.id.webview).getParent();
        FragmentContainerView fragmentContainerView = new FragmentContainerView(bridge.getContext());
        fragmentContainerView.setId(R.id.MediaPlayerFragmentContainerView);
        coordinatorLayout.addView(fragmentContainerView);
    }

    @PluginMethod
    public void create(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "create");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        String url = call.getString("url");
        if (url == null) {
            JSObject ret = new JSObject();
            ret.put("method", "create");
            ret.put("result", false);
            ret.put("message", "Must provide a URL");
            call.resolve(ret);
            return;
        }

        JSObject placementOptions = call.getObject("placement");
        JSObject androidOptions = call.getObject("android");
        JSObject extraOptions = call.getObject("extra");
        JSObject subtitleOptions = extraOptions != null ? extraOptions.getJSObject("subtitles") : null;

        DisplayMetrics metrics = bridge.getContext().getResources().getDisplayMetrics();

        String videoOrientation = placementOptions != null ? placementOptions.getString("videoOrientation", "HORIZONTAL") : "HORIZONTAL";
        String horizontalAlignment = placementOptions != null ? placementOptions.getString("horizontalAlignment", "CENTER") : "CENTER";
        String verticalAlignment = placementOptions != null ? placementOptions.getString("verticalAlignment", "TOP") : "TOP";

        Integer paramHorizontalMargin = placementOptions != null ? placementOptions.getInteger("horizontalMargin", null) : null;
        Integer paramVerticalMargin = placementOptions != null ? placementOptions.getInteger("verticalMargin", null) : null;

        int horizontalMargin = paramHorizontalMargin != null ? (int) (paramHorizontalMargin * metrics.density) : 0;
        int verticalMargin = paramVerticalMargin != null ? (int) (paramVerticalMargin * metrics.density) : 0;

        Integer paramHeight = placementOptions != null ? placementOptions.getInteger("height", null) : null;
        Integer paramWidth = placementOptions != null ? placementOptions.getInteger("width", null) : null;

        int height;
        int width;

        if (paramHeight != null && paramWidth != null) {
            width = (int) (paramWidth * metrics.density);
            height = (int) (paramHeight * metrics.density);
        } else if (paramHeight == null && paramWidth == null) {
            if (Objects.equals(videoOrientation, "HORIZONTAL")) {
                width = metrics.widthPixels;
                height = (int) (width * (new Rational(9, 16).floatValue()));
            } else {
                height = metrics.heightPixels;
                width = (int) (height * (new Rational(9, 16).floatValue()));
            }
        } else if (paramHeight != null) {
            height = (int) (paramHeight * metrics.density);
            if (Objects.equals(videoOrientation, "HORIZONTAL")) {
                width = (int) (paramHeight * metrics.density * (new Rational(16, 9).floatValue()));
            } else {
                width = (int) (paramHeight * metrics.density * (new Rational(9, 16).floatValue()));
            }
        } else {
            width = (int) (paramWidth * metrics.density);
            if (Objects.equals(videoOrientation, "HORIZONTAL")) {
                height = (int) (width * (new Rational(16, 9).floatValue()));
            } else {
                height = (int) (width * (new Rational(9, 16).floatValue()));
            }
        }

        if (width + horizontalMargin > metrics.widthPixels) {
            width = metrics.widthPixels - horizontalMargin;
            if (Objects.equals(videoOrientation, "HORIZONTAL")) {
                height = (int) (width * (new Rational(9, 16).floatValue()));
            } else {
                height = (int) (width * (new Rational(16, 9).floatValue()));
            }
        }
        if (height + verticalMargin > metrics.heightPixels) {
            height = metrics.heightPixels - verticalMargin;
            if (Objects.equals(videoOrientation, "HORIZONTAL")) {
                width = (int) (height * (new Rational(16, 9).floatValue()));
            } else {
                width = (int) (height * (new Rational(9, 16).floatValue()));
            }
        }

        PlacementOptions placement = new PlacementOptions(
            height,
            width,
            videoOrientation,
            horizontalAlignment,
            verticalAlignment,
            horizontalMargin,
            verticalMargin
        );

        AndroidOptions android = new AndroidOptions(
            androidOptions == null || androidOptions.optBoolean("enableChromecast", true),
            androidOptions == null || androidOptions.optBoolean("enablePiP", true),
            androidOptions == null || androidOptions.optBoolean("enableBackgroundPlay", true),
            androidOptions != null && androidOptions.optBoolean("openInFullscreen", false),
            androidOptions != null && androidOptions.optBoolean("automaticallyEnterPiP", false),
            androidOptions == null || androidOptions.optBoolean("fullscreenOnLandscape", true),
            androidOptions == null || androidOptions.optBoolean("stopOnTaskRemoved", false)
        );

        SubtitleOptions subtitles = null;
        if (subtitleOptions != null) {
            double fontSize = Double.parseDouble("12");
            try {
                fontSize = subtitleOptions.getDouble("fontSize");
            } catch (NullPointerException | JSONException ignored) {}
            subtitles = new SubtitleOptions(
                subtitleOptions.getString("url", null),
                subtitleOptions.getString("language", "English"),
                subtitleOptions.getString("foregroundColor", null),
                subtitleOptions.getString("backgroundColor", null),
                fontSize
            );
        }

        double rate = 1;
        try {
            rate = extraOptions.getDouble("rate");
        } catch (NullPointerException | JSONException ignored) {}

        ExtraOptions extra = new ExtraOptions(
            extraOptions != null ? extraOptions.getString("title") : null,
            extraOptions != null ? extraOptions.getString("subtitle") : null,
            extraOptions != null ? extraOptions.getString("poster", null) : null,
            extraOptions != null ? extraOptions.getString("artist", null) : null,
            rate,
            subtitles,
            extraOptions != null && extraOptions.optBoolean("autoPlayWhenReady", false),
            extraOptions != null && extraOptions.optBoolean("loopOnEnd", false),
            extraOptions == null || extraOptions.optBoolean("showControls", true),
            extraOptions != null ? extraOptions.getJSObject("headers") : null
        );
        bridge.getActivity().runOnUiThread(() -> implementation.create(call, playerId, url, placement, android, extra));
    }

    @PluginMethod
    public void play(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "play");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.play(call, playerId));
    }

    @PluginMethod
    public void pause(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "pause");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.pause(call, playerId));
    }

    @PluginMethod
    public void getDuration(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "getDuration");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.getDuration(call, playerId));
    }

    @PluginMethod
    public void getCurrentTime(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "getCurrentTime");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.getCurrentTime(call, playerId));
    }

    @PluginMethod
    public void setCurrentTime(final PluginCall call) {
        String playerId = call.getString("playerId");
        Double time = call.getDouble("time");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "setCurrentTime");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        if (time == null) {
            JSObject ret = new JSObject();
            ret.put("method", "setCurrentTime");
            ret.put("result", false);
            ret.put("message", "Must provide a time");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.setCurrentTime(call, playerId, time.longValue()));
    }

    @PluginMethod
    public void isPlaying(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "isPlaying");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.isPlaying(call, playerId));
    }

    @PluginMethod
    public void isMuted(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "isMuted");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.isMuted(call, playerId));
    }

    @PluginMethod
    public void setVisibilityBackgroundForPiP(final PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("method", "setVisibilityBackgroundForPiP");
        ret.put("result", false);
        ret.put("message", "Method setVisibilityBackgroundForPiP not implemented for Android");
        call.resolve(ret);
    }

    @PluginMethod
    public void mute(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "mute");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.mute(call, playerId));
    }

    @PluginMethod
    public void getVolume(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "getVolume");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.getVolume(call, playerId));
    }

    @PluginMethod
    public void setVolume(final PluginCall call) {
        String playerId = call.getString("playerId");
        Double volume = call.getDouble("volume");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "setVolume");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        if (volume == null) {
            JSObject ret = new JSObject();
            ret.put("method", "setVolume");
            ret.put("result", false);
            ret.put("message", "Must provide a volume");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.setVolume(call, playerId, volume));
    }

    @PluginMethod
    public void getRate(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "getRate");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.getRate(call, playerId));
    }

    @PluginMethod
    public void setRate(final PluginCall call) {
        String playerId = call.getString("playerId");
        Double rate = call.getDouble("rate");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "setRate");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        if (rate == null) {
            JSObject ret = new JSObject();
            ret.put("method", "setRate");
            ret.put("result", false);
            ret.put("message", "Must provide a rate");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.setRate(call, playerId, rate));
    }

    @PluginMethod
    public void remove(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "remove");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.remove(call, playerId));
    }

    @PluginMethod
    public void removeAll(final PluginCall call) {
        bridge.getActivity().runOnUiThread(() -> implementation.removeAll(call));
    }

    @PluginMethod
    public void isFullScreen(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "isFullScreen");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.isFullScreen(call, playerId));
    }

    @PluginMethod
    public void toggleFullScreen(final PluginCall call) {
        String playerId = call.getString("playerId");
        if (playerId == null) {
            JSObject ret = new JSObject();
            ret.put("method", "toggleFullScreen");
            ret.put("result", false);
            ret.put("message", "Must provide a PlayerId");
            call.resolve(ret);
            return;
        }
        bridge.getActivity().runOnUiThread(() -> implementation.toggleFullScreen(call, playerId));
    }
}

File: ios/Sources/MediaPlayerPlugin/MediaPlayerPlugin.swift
import Foundation
import Capacitor
import UIKit
import AVKit

@objc(MediaPlayerPlugin)
public class MediaPlayerPlugin: CAPPlugin, CAPBridgedPlugin {
    
    public let identifier = "MediaPlayerPlugin"
    public let jsName = "MediaPlayer"

    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "create", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "play", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pause", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDuration", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCurrentTime", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setCurrentTime", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isPlaying", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isMuted", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setVisibilityBackgroundForPiP", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "mute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVolume", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setVolume", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getRate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setRate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeAll", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = MediaPlayer()

    var mediaPlayerPlayObserver: Any?;
    var mediaPlayerPauseObserver: Any?;
    var mediaPlayerEndedObserver: Any?;
    var mediaPlayerReadyObserver: Any?;
    var mediaPlayerSeekObserver: Any?
    var mediaPlayerTimeUpdatedObserver: Any?
    var mediaPlayerFullscreenObserver: Any?
    var mediaPlayerPictureInPictureObserver: Any?
    var mediaPlayerIsPlayingInBackgroundObserver: Any?

    override public func load() {
        addNotificationCenterObservers();
        implementation.setBridge(bridge: self.bridge!)
    }

    deinit {
        removeNotificationCenterObservers();
    }

    @objc func create(_ call: CAPPluginCall) {
        guard let playerId = call.options["playerId"] as? String else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "play", "message": error]);
            return
        }
        guard let url = call.options["url"] as? String else {
            let error: String = "Must provide a url"
            print(error);
            call.resolve(["result": false, "method": "play", "message": error]);
            return
        }
        
        let dictUrl: [String: Any] = getURLFromFilePath(filePath: url)
        if let message = dictUrl["message"] as? String {
            if message.count > 0 {
                call.resolve([ "result": false, "method": "create", "message": message])
                return
            }
        }
        guard let parsedUrl = dictUrl["url"] as? URL else {
            call.resolve([ "result": false, "method": "create", "message": "URL not defined"])
            return
        }


        let placementOptions = call.options["placement"] as? [String: Any] ?? [:]
        let iosOptions = call.options["ios"] as? [String: Any] ?? [:]
        let extraOptions = call.options["extra"] as? [String: Any] ?? [:]
        let subtitleOptions = extraOptions["subtitles"] as? [String: Any] ?? nil
        
        let ios = MediaPlayerIosOptions(
            enableExternalPlayback: iosOptions["enableExternalPlayback"] as? Bool, enablePiP: iosOptions["enablePiP"] as? Bool, enableBackgroundPlay: iosOptions["enableBackgroundPlay"] as? Bool, openInFullscreen: iosOptions["openInFullscreen"] as? Bool, automaticallyEnterPiP: iosOptions["automaticallyEnterPiP"] as? Bool, automaticallyHideBackgroundForPip: iosOptions["automaticallyHideBackgroundForPip"] as? Bool, fullscreenOnLandscape: iosOptions["fullscreenOnLandscape"] as? Bool, allowsVideoFrameAnalysis: iosOptions["allowsVideoFrameAnalysis"] as? Bool
        )
                        
        var subTitle: URL?
        if subtitleOptions != nil {
            let dictSubTitle: [String: Any] = getURLFromFilePath(filePath: subtitleOptions!["url"] as! String)
            if let message = dictSubTitle["message"] as? String {
                if message.count > 0 {
                    call.resolve([ "result": false, "method": "create",
                                   "message": message])
                    return
                }
            }
            guard let sturl = dictSubTitle["url"] as? URL else {
                call.resolve([ "result": false, "method": "create",
                               "message": "Subtitles URL not defined"])
                return
            }
            subTitle = sturl
        }
        
        let subtitles = (subtitleOptions != nil) ? MediaPlayerSubtitleOptions(
            url: subTitle!, language: subtitleOptions?["language"] as? String, foregroundColor: subtitleOptions!["foregroundColor"] as? String, backgroundColor: subtitleOptions!["backgroundColor"] as? String, fontSize:subtitleOptions?["fontSize"] as? Float
        ) : nil

        var posterURL: URL?
        if let poster = extraOptions["poster"] as? String {
            let posterDictUrl: [String: Any] = getURLFromFilePath(filePath: poster)
            if let parsedUrl = posterDictUrl["url"] as? URL {
                posterURL = parsedUrl
            }
        }
        
        let extra = MediaPlayerExtraOptions(
            title: extraOptions["title"] as? String, subtitle: extraOptions["subtitle"] as? String, poster: posterURL, artist: extraOptions["artist"] as? String, rate: extraOptions["rate"] as? Float, subtitles: subtitles, autoPlayWhenReady: extraOptions["autoPlayWhenReady"] as? Bool, loopOnEnd: extraOptions["loopOnEnd"] as? Bool, showControls:extraOptions["showControls"] as? Bool, headers: extraOptions["headers"] as? [String: String]
        )

        let placement = MediaPlayerPlacementOptions(
            height: placementOptions["height"] as? Float, width: placementOptions["width"] as? Float, videoOrientation: placementOptions["videoOrientation"] as? String,
            verticalMargin: placementOptions["verticalMargin"] as? Float, horizontalMargin: placementOptions["horizontalMargin"] as? Float, horizontalAlignment: placementOptions["horizontalAlignment"] as? String, verticalAlignment: placementOptions["verticalAlignment"] as? String
        )
        
        implementation.create(call: call, playerId: playerId, url: parsedUrl, placement: placement, ios: ios, extra: extra)
    }

    @objc func play(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "play", "message": error]);
            return
        }
        implementation.play(call: call, playerId: playerId)
    }

    @objc func pause(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "pause", "message": error]);
            return
        }
        implementation.pause(call: call, playerId: playerId)
    }

    @objc func getDuration(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "getDuration", "message": error]);
            return
        }
        implementation.getDuration(call: call, playerId: playerId)
    }

    @objc func getCurrentTime(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "getCurrentTime", "message": error]);
            return
        }
        implementation.getCurrentTime(call: call, playerId: playerId);
    }

    @objc func setCurrentTime(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "setCurrentTime", "message": error]);
            return
        }
        guard let time = call.getDouble("time") else {
            let error: String = "Must provide a time"
            print(error);
            call.resolve(["result": false, "method": "setCurrentTime", "message": error]);
            return
        }
        implementation.setCurrentTime(call: call, playerId: playerId, time: time);
    }

    @objc func isPlaying(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "isPlaying", "message": error]);
            return
        }
        implementation.isPlaying(call: call, playerId: playerId);
    }

    @objc func isMuted(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "isMuted", "message": error]);
            return
        }
        implementation.isMuted(call: call, playerId: playerId)
    }
    
    @objc func setVisibilityBackgroundForPiP(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "setVisibilityBackgroundForPiP", "message": error]);
            return
        }
        guard let isVisible = call.getBool("isVisible") else {
            let error: String = "Must provide isVisible"
            print(error);
            call.resolve(["result": false, "method": "setVisibilityBackgroundForPiP", "message": error]);
            return
        }
        implementation.setVisibilityBackgroundForPiP(call: call, playerId: playerId, isVisible: isVisible)
    }

    @objc func mute(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "mute", "message": error]);
            return
        }
        implementation.mute(call: call, playerId: playerId)
    }

    @objc func getVolume(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "getVolume", "message": error]);
            return
        }
        implementation.getVolume(call: call, playerId: playerId)
    }

    @objc func setVolume(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "setVolume", "message": error]);
            return
        }
        guard let volume = call.getFloat("volume") else {
            let error: String = "Must provide a volume"
            print(error);
            call.resolve(["result": false, "method": "setVolume", "message": error]);
            return
        }
        implementation.setVolume(call: call, playerId: playerId, volume: volume)
    }

    @objc func getRate(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "getRate", "message": error]);
            return
        }
        implementation.getRate(call: call, playerId: playerId)
    }

    @objc func setRate(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "setRate", "message": error]);
            return
        }
        guard let rate = call.getFloat("rate") else {
            let error: String = "Must provide a rate"
            print(error);
            call.resolve(["result": false, "method": "setRate", "message": error]);
            return
        }
        implementation.setRate(call: call, playerId: playerId, rate: rate)
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let playerId = call.getString("playerId") else {
            let error: String = "Must provide a playerId"
            print(error);
            call.resolve(["result": false, "method": "remove", "message": error]);
            return
        }
        implementation.remove(call: call, playerId: playerId)
    }

    @objc func removeAll(_ call: CAPPluginCall) {
        implementation.removeAll(call: call)
    }

}
